jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Connection.java

```diff
@@ -290,4 +332,143 @@ public InvocationType getInvocationType()
             return InvocationType.EITHER;
         }
     }
+
+    private class ParserListener implements Parser.Listener
+    {
+        private final Parser.Listener listener;
+
+        private ParserListener(Parser.Listener listener)
+        {
+            this.listener = listener;
+        }
+
+        @Override
+        public void onData(DataFrame frame)
+        {
+            NetworkBuffer buffer = producer.buffer;
+            buffer.retain();
+            // TODO: remove downcast.
+            ((HTTP2Session)session).onData(frame, buffer);
+        }
+
+        @Override
+        public void onHeaders(HeadersFrame frame)
+        {
+            listener.onHeaders(frame);
+        }
+
+        @Override
+        public void onPriority(PriorityFrame frame)
+        {
+            listener.onPriority(frame);
+        }
+
+        @Override
+        public void onReset(ResetFrame frame)
+        {
+            listener.onReset(frame);
+        }
+
+        @Override
+        public void onSettings(SettingsFrame frame)
+        {
+            listener.onSettings(frame);
+        }
+
+        @Override
+        public void onPushPromise(PushPromiseFrame frame)
+        {
+            listener.onPushPromise(frame);
+        }
+
+        @Override
+        public void onPing(PingFrame frame)
+        {
+            listener.onPing(frame);
+        }
+
+        @Override
+        public void onGoAway(GoAwayFrame frame)
+        {
+            listener.onGoAway(frame);
+        }
+
+        @Override
+        public void onWindowUpdate(WindowUpdateFrame frame)
+        {
+            listener.onWindowUpdate(frame);
+        }
+
+        @Override
+        public void onConnectionFailure(int error, String reason)
+        {
+            listener.onConnectionFailure(error, reason);
+        }
+    }
+
+    private class NetworkBuffer implements Callback, Retainable
```

Comment:
I'm not sure I like this being a Callback, specially as we will eventually split release semantic away from succeeded!